File: java/src/test/java/io/github/xesam/linkseal/VariableResolverTest.java
package io.github.xesam.linkseal;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.github.xesam.linkseal.resolver.VariableResolver;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

final class VariableResolverTest {

  @Test
  void hasNoValuesBeforeAnyAreRegistered() {
    assertNull(new VariableResolver().resolveOne("userId"));
  }

  @Test
  void resolvesRegisteredValuesByPlaceholderName() {
    VariableResolver resolver = new VariableResolver();
    resolver.setVariables(Map.of("userId", "user_123", "token", "sess_abc"));
    assertEquals("user_123", resolver.resolveOne("userId"));
    assertEquals("sess_abc", resolver.resolveOne("token"));
  }

  @Test
  void returnsNullForUnregisteredName() {
    VariableResolver resolver = new VariableResolver();
    resolver.setVariables(Map.of("userId", "user_123"));
    assertNull(resolver.resolveOne("token"));
  }

  @Test
  void setVariablesMergesInsteadOfReplacing() {
    VariableResolver resolver = new VariableResolver();
    resolver.setVariables(Map.of("userId", "user_123"));
    resolver.setVariables(Map.of("token", "sess_abc"));
    assertEquals("user_123", resolver.resolveOne("userId"));
    assertEquals("sess_abc", resolver.resolveOne("token"));
  }

  @Test
  void setVariablesOverridesExistingValue() {
    VariableResolver resolver = new VariableResolver();
    resolver.setVariables(Map.of("userId", "first"));
    resolver.setVariables(Map.of("userId", "second"));
    assertEquals("second", resolver.resolveOne("userId"));
  }

  @Test
  void fallsBackToProviderWhenNameIsNotRegistered() {
    VariableResolver resolver = new VariableResolver();
    resolver.setResolver(name -> "token".equals(name) ? "fresh_token" : null);
    assertEquals("fresh_token", resolver.resolveOne("token"));
  }

  /** Registered values win, and the provider must not even be consulted for them. */
  @Test
  void prefersRegisteredValuesOverProvider() {
    List<String> calls = new ArrayList<>();
    VariableResolver resolver = new VariableResolver();
    resolver.setVariables(Map.of("userId", "user_123"));
    resolver.setResolver(name -> {
      calls.add(name);
      return "FROM_PROVIDER";
    });

    assertEquals("user_123", resolver.resolveOne("userId"));
    assertTrue(calls.isEmpty());
  }

  @Test
  void consultsProviderOnlyForNamesTheMapMisses() {
    List<String> calls = new ArrayList<>();
    VariableResolver resolver = new VariableResolver();
    resolver.setVariables(Map.of("userId", "user_123"));
    resolver.setResolver(name -> {
      calls.add(name);
      return "token".equals(name) ? "fresh" : null;
    });

    resolver.resolve(List.of("userId", "token"));

    assertEquals(List.of("token"), calls);
  }

  @Test
  void treatsProviderReturningNullAsMissing() {
    VariableResolver resolver = new VariableResolver();
    resolver.setResolver(name -> null);
    assertNull(resolver.resolveOne("token"));
  }

  @Test
  void setResolverNullClearsProvider() {
    VariableResolver resolver = new VariableResolver();
    resolver.setResolver(name -> "value");
    resolver.setResolver(null);
    assertNull(resolver.resolveOne("token"));
  }

  /** Unresolved names must survive as null, or PolicyProcessor cannot see them. */
  @Test
  void resolveKeepsUnresolvedNamesAsNull() {
    VariableResolver resolver = new VariableResolver();
    resolver.setVariables(Map.of("userId", "user_123"));

    Map<String, String> expected = new LinkedHashMap<>();
    expected.put("userId", "user_123");
    expected.put("token", null);
    expected.put("locale", null);

    assertEquals(expected, resolver.resolve(List.of("userId", "token", "locale")));
  }

  @Test
  void resolveReturnsEntryForEveryRequestedName() {
    Map<String, String> result = new VariableResolver().resolve(List.of("a", "b"));
    assertEquals(List.of("a", "b"), new ArrayList<>(result.keySet()));
  }

  @Test
  void resolveOnEmptyNameListReturnsEmptyMap() {
    assertEquals(Map.of(), new VariableResolver().resolve(List.of()));
  }
}

File: javascript/packages/unilink-seal-resolver/src/__tests__/VariableResolver.test.ts
import { describe, it, expect, vi } from 'vitest';
import { VariableResolver } from '../VariableResolver.js';

describe('VariableResolver', () => {
  it('has no values before any are registered', () => {
    const resolver = new VariableResolver();
    expect(resolver.resolveOne('userId')).toBeNull();
  });

  it('resolves registered values by placeholder name', () => {
    const resolver = new VariableResolver();
    resolver.setVariables({ userId: 'user_123', token: 'sess_abc' });
    expect(resolver.resolveOne('userId')).toBe('user_123');
    expect(resolver.resolveOne('token')).toBe('sess_abc');
  });

  it('returns null for an unregistered name', () => {
    const resolver = new VariableResolver();
    resolver.setVariables({ userId: 'user_123' });
    expect(resolver.resolveOne('token')).toBeNull();
  });

  it('setVariables merges instead of replacing', () => {
    const resolver = new VariableResolver();
    resolver.setVariables({ userId: 'user_123' });
    resolver.setVariables({ token: 'sess_abc' });
    expect(resolver.resolveOne('userId')).toBe('user_123');
    expect(resolver.resolveOne('token')).toBe('sess_abc');
  });

  it('setVariables overrides an existing value', () => {
    const resolver = new VariableResolver();
    resolver.setVariables({ userId: 'first' });
    resolver.setVariables({ userId: 'second' });
    expect(resolver.resolveOne('userId')).toBe('second');
  });

  it('falls back to the provider when a name is not registered', () => {
    const resolver = new VariableResolver();
    resolver.setResolver((name) => (name === 'token' ? 'fresh_token' : null));
    expect(resolver.resolveOne('token')).toBe('fresh_token');
  });

  // Registered values win, and the provider must not even be consulted for them.
  it('prefers registered values over the provider', () => {
    const provider = vi.fn(() => 'FROM_PROVIDER');
    const resolver = new VariableResolver();
    resolver.setVariables({ userId: 'user_123' });
    resolver.setResolver(provider);

    expect(resolver.resolveOne('userId')).toBe('user_123');
    expect(provider).not.toHaveBeenCalled();
  });

  it('consults the provider only for names the map misses', () => {
    const provider = vi.fn((name: string) => (name === 'token' ? 'fresh' : null));
    const resolver = new VariableResolver();
    resolver.setVariables({ userId: 'user_123' });
    resolver.setResolver(provider);

    resolver.resolve(['userId', 'token']);

    expect(provider).toHaveBeenCalledTimes(1);
    expect(provider).toHaveBeenCalledWith('token');
  });

  it('treats a provider returning null as missing', () => {
    const resolver = new VariableResolver();
    resolver.setResolver(() => null);
    expect(resolver.resolveOne('token')).toBeNull();
  });

  it('setResolver(null) clears a previously registered provider', () => {
    const resolver = new VariableResolver();
    resolver.setResolver(() => 'value');
    resolver.setResolver(null);
    expect(resolver.resolveOne('token')).toBeNull();
  });

  // Unresolved names must survive as null, or PolicyProcessor cannot see them.
  it('resolve keeps unresolved names as null', () => {
    const resolver = new VariableResolver();
    resolver.setVariables({ userId: 'user_123' });

    expect(resolver.resolve(['userId', 'token', 'locale'])).toEqual({
      userId: 'user_123',
      token: null,
      locale: null,
    });
  });

  it('resolve returns an entry for every requested name', () => {
    const resolver = new VariableResolver();
    const result = resolver.resolve(['a', 'b']);
    expect(Object.keys(result)).toEqual(['a', 'b']);
  });

  it('resolve on an empty name list returns an empty map', () => {
    expect(new VariableResolver().resolve([])).toEqual({});
  });
});
